language/move-vm/runtime/src/loader.rs

```diff
@@ -690,32 +742,69 @@ impl<'a> Resolver<'a> {
         Self { loader, binary }
     }
 
+    //
+    // Constant resolution
+    //
+
     pub(crate) fn constant_at(&self, idx: ConstantPoolIndex) -> &Constant {
         match &self.binary {
             BinaryType::Module(module) => module.module.constant_at(idx),
             BinaryType::Script(script) => script.script.constant_at(idx),
         }
     }
 
-    pub(crate) fn function_at(&self, idx: FunctionHandleIndex) -> Arc<Function> {
+    //
+    // Function resolution
+    //
+
+    pub(crate) fn function_from_handle(&self, idx: FunctionHandleIndex) -> Arc<Function> {
         let idx = match &self.binary {
             BinaryType::Module(module) => module.function_at(idx.0),
             BinaryType::Script(script) => script.function_at(idx.0),
         };
         self.loader.function_at(idx)
     }
 
-    pub(crate) fn function_instantiation_at(
+    pub(crate) fn function_from_generic(&self, idx: FunctionInstantiationIndex) -> Arc<Function> {
+        let func_inst = match &self.binary {
+            BinaryType::Module(module) => module.function_instantiation_at(idx.0),
+            BinaryType::Script(script) => script.function_instantiation_at(idx.0),
+        };
+        self.loader.function_at(func_inst.handle)
+    }
+
+    pub(crate) fn materialize_generic_function(
         &self,
         idx: FunctionInstantiationIndex,
-    ) -> &FunctionInstantiation {
-        match &self.binary {
+        type_params: &[Type],
+    ) -> VMResult<Vec<Type>> {
+        let func_inst = match &self.binary {
             BinaryType::Module(module) => module.function_instantiation_at(idx.0),
             BinaryType::Script(script) => script.function_instantiation_at(idx.0),
+        };
+        let mut instantiation = vec![];
+        for ty in &func_inst.instantiation {
+            instantiation.push(ty.subst(type_params)?);
         }
+        Ok(instantiation)
+    }
+
+    pub(crate) fn generic_function_arity(
```

Comment:
I found this name a bit confusing. Upon first glance, "arity" could mean a few things to me:
1. The total number of parameters, including type parameters and value parameters
2. The number of value parameters.
3. The number of type parameters.
and it was only clear until I read the body of the function.

I'd personally prefer a more straightforward name including sth like "num_type_params".

Reply:
went for `type_params_count`